### ISSUE TITLE
How to use this library

### ISSUE THREAD

**Issue by irtusb** [vurtun/nuklear#226]
(in progress)
### Pre
As far as I understand you need to:
- get nuklear.h somehow (clone, download, copy contents from browser)
- get nuklear implementation compiled by setting `#define NK_IMPLEMENTATION` somewhere and before `#include nuklear.h` (two lines .c file, your main.c file, compiler -D option)
- write a renderer backend or use one of the examples from `demo` subdirectory (sdl, gdi, x11, ...) that can actually draw on screen AND process input (mouse, keyboard, ...)

if you write a custom backend (glue code) there are several paths you can take regarding each area (drawing, font handling, event handling, ...)

**Comment by irtusb** [comment]
Please @vurtun review this if you have time

**Comment by vurtun** [comment]
Thanks for the write up, I probably will spend some time on the weekend and do a review and add some bits and pieces and further correct some information.

**Comment by JesseRMeyer** [comment]
Thanks for writing this up! It's been a good time saver for me.

**Comment by jpcima** [comment]
I am in trouble understanding how the various layouts work and I would be happy to read any documentation on that.
I try to program a GUI where I would like fixed positions for about everything, and I do not find a way to accomplish that. I wish I could just designate a rectangle, put the next widget into it, and repeat.
I it possible to do without layouts altogether?
(I have no idea where to ask except here sorry)

**Comment by vurtun** [comment]
@jpcima do you mean `nk_space_begin`, `nk_space_push` and `nk_space_end`?

**Comment by jpcima** [comment]
I don't understand exactly the functionality of `nk_layout_space_begin`, in particular the purpose of the arguments height and widget_count. However I found I could accomplish exactly what I want using `nk_layout_space_begin(ctx, NK_STATIC, 0, INT_MAX)`.
I think I prefer the same way of positioning the JUCE library does, it is simpler code to me and less headache.

**Comment by vurtun** [comment]
> I think I prefer the same way of positioning the JUCE library does, it is simpler code to me and less headache.

If you have an idea or proposal to enhance nuklear go for it. At least tell me what and how they are doing it better and therefor make nuklear better. But please don't shit in my face, by basically telling me my implementation sucks.

**Comment by jpcima** [comment]
Ok vurtun, sorry if my previous message sound to you this way. But please believe it is not my intent to throw criticism at your quality of work, please do not take my message wrongly. I would not have interest in contributing if nuklear was bad.

Allow me just to clear some things up. I try to make my first GUI and I am in moderately good progress, with my first 2 widgets made. I get most information through my understanding of source code.
Since I have honestly the greatest trouble to understand the layouts, I choose to go with fixed positions, especially because this is what I used in past experience.

I do not suggest to change nuklear. I have little understanding, and no authority to say it should be like this or that. I try to apply a paradigm I am familiar with, and to accomplish that was the objective of my question.
The "JUCE way" is only to be understood as fixed + a fancy set of methods to successively subdivide rectangles.

Apparently the documentation currently lacks in nuklear. I have found this [page](http://wiki.call-cc.org/eggref/4/nuklear) helpful, which is related to Chicken Scheme. If there is a more complete source of information I ignore about, please let me know.

**Comment by dumblob** [comment]
>The "JUCE way" is only to be understood as fixed + a fancy set of methods to successively subdivide rectangles.

I've tried to find any documentation describing positioning and/or layouting in JUCE for about an **hour** yesterday and completely failed, because I did find literally nothing. @jpcima could you please point me to a documentation or source code describing the "JUCE way" to look at?

By the way, if you're interested in automated screen size independent layouting, go and check the comment on Cassowary in https://github.com/vurtun/nuklear/issues/123#issuecomment-260666623 .

**Comment by jpcima** [comment]
> @jpcima could you please point me to a documentation or source code describing the "JUCE way" to look at?

Yes the JUCE author has a [talk](https://www.youtube.com/watch?v=xsCZoE1s_uw) where he briefly explains it, which is you can hear at 7:30-12:00. This is what I was refering to.
I have figured how to do the similar think in nuklear and it works, so no problem.

> By the way, if you're interested in automated screen size independent layouting, go and check the comment on Cassowary in #123 (comment) .

It is not an issue which is on my mind currently but thanks for mentioning.

**Comment by dumblob** [comment]
@jpcima thanks for the link. This method of "chopping rectangles off from the original big canvas by dividing it horizontally or vertically" seems very easy to use. One could say it produces a binary tree of rectangles in a way as is used e.g. in [herbstluftwm](https://www.herbstluftwm.org/ ) for window layouting.

How does one solve (easily) the following things using the "chopping method"? Are there also some "tricks"?

1. Unlimited overlapping (of widgets, backgrounds, etc.). I can imagine one "resets" the coordinates and size of the current state (which holds basically just the coordinates and size of the remaining rectangle) and draws again. But is it really this easy? Could one use it to draw e.g. menus and submenus (recursively) easily?
1. Dynamic screen size, resolution, and DPI changes (e.g. rotation of the viewport as basically all smartphones do; or connecting to a different LCD, etc.).
1. From your experience, the inevitable "layering" (refer to the first question) is how much inconvenient?

>I have figured how to do the similar think in nuklear and it works, so no problem.

Please post your results somewhere. Actually this "chopping" method should be trivial to implement with Nuklear.

@vurtun it's worth mentioning in README as this method is API agnostic and serves as a very good start for any GUI (it uses the "immediate mode" principle *one after another and nothing beforehand* unlike explicit "layouting" or boxing or similar retained UI concepts). We could also try to rewrite some of the examples we have currently to use this method to see how does it scale to larger number of widgets and complexity.

**Comment by jpcima** [comment]
> How does one solve (easily) the following things using the "chopping method"? Are there also some "tricks"?

No particular trick worth noting comes to my mind. I am not sure I am too well placed to answer these questions, because I am a simple user with simple needs, but anyway:
1. > is it really this easy?

My opinion may be naive, but I think so. In composite UIs the overlap is traditionally resolved with z-order attributes, and Imgui makes that z-order explicit in control flow. I do not see a particular problem with this.

2. > Dynamic screen size

If you look back at this talk I linked to, you can observe the positioning is performed in a resize callback, exactly for the purpose to adapt to size changes. It is pretty straight forward. As for anything regarding DPI, I have no idea.

3. > the inevitable "layering" (refer to the first question) is how much inconvenient?

I don't think it is, if you think of something specific please elaborate.

> Please post your results somewhere. Actually this "chopping" method should be trivial to implement with Nuklear.

Nothing very extravagant at the time being. This is the UI of a music intrument plugin.
Since I have a rough but working GUI atm, I focus temporarily to link the interaction to the program logic.
I will later introduce the other music related elements; radial controls, vumeters, plots.. [Here is partial C++ source code.](https://github.com/vurtun/nuklear/files/821701/ui.tar.gz).

![screenshot](https://cloud.githubusercontent.com/assets/17614485/23616888/8020d572-028b-11e7-8ff8-cc9d5519df1d.png)

**Comment by carmel4a** [comment]
Ok, as I don't see better place... As I understand:
```
const struct nk_command *cmd = 0;
nk_foreach(cmd, this->ctx_ptr){
    switch (cmd->type) {
        case NK_COMMAND_NOP:
            break;
        case NK_COMMAND_SCISSOR:
            [HERE]
        case NK_COMMAND_LINE:
            [HERE]
            cmd->sth //doesn't work
        ...
```
in **[HERE]** I should put my render backend. Should I be able to get command properties at this point? If no - how that should works? How can I know exacly which line I need to draw, etc.

**Comment by MrDChristop** [comment]
This text should be in the Wiki.....

**Comment by irtusb** [comment]
> This text should be in the Wiki.....

it was written before anything apart from the readme and examples existed

**Comment by omricarmi** [comment]
did someone implement this using OpenCV?
